Create LookupEntityExtractor.py

JiteshGaikwad opened the pull request:
https://github.com/RasaHQ/rasa/issues/5957#issuecomment-654088366

Lookup Entity Extractor built for Rasa for extracting known entities from a list of examples without the need for training on every set of examples. The extractor is not machine-learned but uses text-matching logic for entity prediction. It doesn't discover any additional values other than the provided list of examples.

Lookup Entity Extractor is a good fit when you have list of exampless that are known and can be expanded as required.

[README.md](https://github.com/JiteshGaikwad/LookupEntityExtractor/blob/master/README.md)

**Proposed changes**:
- ...

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=6183) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=6183) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=6183) it.</sub>

sara-tagger commented:
Thanks for submitting a pull request 🚀 @TyDunn will take a look at it as soon as possible ✨

koaning commented on rasa/nlu/extractors/LookupEntityExtractor.py:
```diff
@@ -0,0 +1,129 @@
+        and check whether it exists in the user's message
+        """
+        results = []
+        for i in range(0, len(entities)):
```
You can replace `range(0, len(entities))` with `range(len(entities))`.

koaning commented on rasa/nlu/extractors/LookupEntityExtractor.py:
```diff
@@ -0,0 +1,129 @@
+
+class LookupEntityExtractor(EntityExtractor):
+    """
+    Searches for entities in the user's message from a list of examples.\n
```
I've noticed the `\n` character appear in the docstrings. Is there a reason for this?

JiteshGaikwad replied:
I have added it so that whenever I see the class definition it would look properly formatted

![image](https://user-images.githubusercontent.com/49223934/87157688-3c63f980-c2dc-11ea-8401-d3fe303df835.png)

Let me know if I need to remove the formatting with "\n"

koaning replied:
Every editor is bound to render it differently, but it's probably best to stick to the PEP guidelines here which is to remove the `\n`.

koaning commented on rasa/nlu/extractors/LookupEntityExtractor.py:
```diff
@@ -0,0 +1,129 @@
+            examples = f.readlines()
+
+            for example in examples:
+                if example.lower().strip() in user_input.lower():
```
You seem to lower case everything here. I can certainly see that there is merit to this but I wonder if this is the behavior you'd want in all scenarios. It might be better to have this be a configuration. @tabergma you might also have an opinion on this.

JiteshGaikwad replied:
If in the Lookup file, if there is any value which is lower case or user query has both the things, so to avoid the casing issue, I preferred lower case everything here. let me know if this behavior increases the code processing time.

koaning replied:
I'll resolve this conversation since tanja replied with a more indepth correspondence below.

koaning commented on rasa/nlu/extractors/LookupEntityExtractor.py:
```diff
@@ -0,0 +1,129 @@
+                if example.lower().strip() in user_input.lower():
+                    start_index = user_input.lower().index(example.lower().strip())
+                    end_index = start_index + len(example.strip())
+                    temp = {}
```
This is personal preference, but it might be appropriate to declare the dictionary directly. 

```python
temp = {
  "entity": entities[i], 
  "start": start_index,
  "end": end_index,
  "value": user_input[start_index:end_index]
}
```

JiteshGaikwad replied:
Updated the code to declare the dictionary directly.

koaning commented on rasa/nlu/extractors/LookupEntityExtractor.py:
```diff
@@ -0,0 +1,129 @@
+        """
+        results = []
+        for i in range(0, len(entities)):
+            f = open(file_path[i], "r")
```
It might be preferable to explicitly use a context manager here. Something like;

```python
with open(file_path[i], "r") as f:
    ...
```

Another alternative might be to use pathlib. 

```python
import pathlib
pathlib.Path(file_path[i]).read_text()
```

For details on why there's a link to [stackoverflow](https://stackoverflow.com/questions/7395542/is-explicitly-closing-files-important).

JiteshGaikwad replied:
@koaning how about `f.close()` won't it close the file connection?

koaning commented:
Hi @JiteshGaikwad thanks for the PR! I've given it a first order review, mostly some style things for now. Feel free to discuss them if you're unsure if my suggestion is valid. 

One thing that is currently missing is a set of unit tests though. Have you thought about how you'd want to unit test this component? Feel free to ask for help if this is unfamiliar territory.

JiteshGaikwad commented:
> Hi @JiteshGaikwad thanks for the PR! I've given it a first-order review, mostly some style things for now. Feel free to discuss them if you're unsure if my suggestion is valid.
> 
> One thing that is currently missing is a set of unit tests though. Have you thought about how you'd want to unit test this component? Feel free to ask for help if this is unfamiliar territory.

@koaning 
I haven't used the unit testing yet for my previous work 😅 ya which is unfamiliar territory for me.

Thanks

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,129 @@
+                    )
+                    break
+
+    def add_extractor_name(
```
You don't need this method here. It is already defined in the parent class.

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,129 @@
+
+    @staticmethod
+    def _parse_entities(self, text: Text) -> List[Dict[Text, Any]]:
+        """
+        pass the user input to  extract the entities
+
+        """
```
```suggestion
        """Extract entities from the user input."""
```

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,129 @@
+    """
+
+    defaults = {
+        "entities": None,
+        "files_path": None
```
Can you please add comments what those values are for? Thanks.

tabergma replied:
If I understand it correctly `entities` and `files_path` are both lists. So the first file path lists all values of the first entity type listed in `entities`. Maybe it would be simple to store this as a dict. For example
```
- name: LookupEntityExtractor
   lookup:
     city: /some/path/city.txt
     person: /some/other/path/person.txt
```
What do you think?

tabergma replied:
If you store them as dict, it would also simplify the checks in `__init__` quite a bit.

JiteshGaikwad replied:
@tabergma thanks for the review, I will make those changes

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,129 @@
+                    results.append(temp)
+        return results
+
+    def process(self, message, **kwargs):
```
```suggestion
    def process(self, message: Message, **kwargs: Any) -> None:
```

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+                Please configure the lookup entities in the config.yml."
+            )
+
+    def add_entities_if_lookup_entity(self, entity, file_path) -> None:
```
```suggestion
    def _validate_lookup_entry(self, entity: Text, file_path: Text) -> None:
```

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+
+    def add_entities_if_lookup_entity(self, entity, file_path) -> None:
+        if file_path is not None:
+            if os.path.isfile(file_path):
+                pass
+            else:
+                # remove the entity from the lookup dictionary,
+                # if the file path doesn't exist
+                self.component_config["lookup"].pop(entity)
+                common_utils.raise_warning(
+                    f"can't extract lookup entity: '{entity}',\
+                    make sure the provided file '{file_path}' exists.")
```
```suggestion
            if not os.path.isfile(file_path):
                # remove the entity from the lookup dictionary,
                # if the file path doesn't exist
                self.component_config["lookup"].pop(entity)
                common_utils.raise_warning(
                    f"The file path '{file_path}' for entity '{entity}' does not exist. Please provide a valid file path.")
```

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+            # remove the entity from the lookup dictionary,
+            # if the file path is not provided
+            self.component_config["lookup"].pop(entity)
+            common_utils.raise_warning(
+                f"can't extract '{entity}' entity,\
+                please provide the example file.")
```
```suggestion
                common_utils.raise_warning(
                    f"No file path for entity '{entity}' was given. Please provide a valid file path.")
```

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+                # the lookup entities, assign it None value
+                self.component_config["lookup"] = None
+        else:
+            common_utils.raise_warning(
+                "Can't extract Lookup Entities,\
+                Please configure the lookup entities in the config.yml."
+            )
```
I would actually raise a `ValueError` here. If no lookup dictionary is configured, the component cannot be used at all. This should result in a failure.

tabergma replied:
Same for the other cases above. E.g. if all entries in the lookup dictionary are invalid, so that nothing is left, raise a `ValueError`.

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+
+    def _parse_entities(self, user_input: Text) -> List[Dict[Text, Any]]:
+        """Extract entities from the user input."""
+        if self.component_config["lookup"] is not None:
```
If we raise a `ValueError` in case the lookup dict is not set, we can also remove this check as we know that it exists and that it is valid.

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+    def _parse_entities(self, user_input: Text) -> List[Dict[Text, Any]]:
+        """Extract entities from the user input."""
+        if self.component_config["lookup"] is not None:
+            for entity, file_path in list(
+                    self.component_config["lookup"].items()):
```
```suggestion
            for entity, file_path in self.component_config["lookup"].items():
```
I don't think you need to wrap this in a list.

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+            return []
+
+    @staticmethod
+    def _parse_all_entities(
+        user_input: str, entity: list, file_path: list
+    ) -> List[Dict[Text, Any]]:
```
```suggestion
    def _extract_entities(user_input: Text, entity: Text, file_path: Text) -> List[Dict[Text, Any]]:
```

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+            examples = f.readlines()
+
+            for example in examples:
+                example = example.lower().strip()
```
Let's add an option `lowercase`. We can set its default value to `True`. But I think it might happen that users need to consider the casing when using this component.

tabergma commented on rasa/nlu/extractors/lookup_entity_extractor.py:
```diff
@@ -0,0 +1,119 @@
+        """
+        results = []
+        with open(file_path, "r") as f:
+            examples = f.readlines()
+
+            for example in examples:
```
Let's read the file line by line to avoid memory issues.
```
        with open(file_path, "r") as file:
            for example in file:
```

koaning commented:
@tabergma as far as unit tests go, I'm wondering what items we'd really like to test. From the top of my head I'll list a few here so we can discuss then. 

1. Make sure we properly add the entities.
2. Make sure the lowercase setting works as expected.
3. Make sure we don't remove previous entities? 

I'm also thinking about documentation. Do we want to add an entry to the .rst documents right away? I can help out with this if it is appreciated.

tabergma commented:
@koaning The tests you suggested sound good. 
> Make sure we don't remove previous entities?

This is a good point. We should definitely not overwrite existing entities.

Regarding documentation: We should not merge without at least an entry in the `component.rst` page. There are some other pages, which we could update, but this page is the most important one to update, I would say.

JiteshGaikwad commented:
@tabergma I have added the suggested changes, also I have added the checks for the casing.
@koaning need help in understanding point 1 and 3 below:
> 1. Make sure we properly add the entities.
> 2. Make sure the lowercase setting works as expected.
> 3. Make sure we don't remove previous entities?

Thanks.

tabergma commented:
Thanks again @JiteshGaikwad. We added test and documentation. Here is the link to the final PR: https://github.com/RasaHQ/rasa/pull/6214/files

JiteshGaikwad commented:
Thanks @tabergma  and @koaning
